[composer] drag-and-drop between spaces is broken for sketches

Steps to reproduce:

1. Create two spaces
2. Create a Sketch in the first
3. drag-and-drop the sketch into the second space

https://github.com/dxos/dxos/assets/1310143/38416b19-e531-4700-b098-37556ca89d4a

Console error:
```
react-D-u6Bfzb.js:32 Error: invariant violation [isReactiveObject(obj)] at packages/core/echo/echo-db/src/automerge/automerge-object.ts:16
```

--- a/packages/apps/types/src/serializer/clone.ts
+++ b/packages/apps/types/src/serializer/clone.ts
@@ -18,5 +18,5 @@ export const cloneObject = async (object: Expando): Promise<Expando> => {
   const serializer = serializers[typename] ?? serializers.default;
   invariant(serializer, `No serializer for type: ${typename}`);
   const content = await serializer.serialize({ object, serializers });
-  return serializer.deserialize({ content, serializers, preserveId: false });
+  return serializer.deserialize({ content, serializers, newId: true });
 };

--- a/packages/apps/types/src/serializer/clone.ts
+++ b/packages/apps/types/src/serializer/clone.ts
@@ -11,12 +11,12 @@ import { TypeOfExpando } from './types';
 /**
  * @deprecated Workaround for ECHO not supporting clone.
  */
-// TODO(burdon): Remove?
+// TODO(burdon): Remove.
 export const cloneObject = async (object: Expando): Promise<Expando> => {
   const schema = getSchema(object);
   const typename = schema ? getEchoObjectAnnotation(schema)?.typename ?? TypeOfExpando : TypeOfExpando;
   const serializer = serializers[typename] ?? serializers.default;
   invariant(serializer, `No serializer for type: ${typename}`);
   const content = await serializer.serialize({ object, serializers });
-  return serializer.deserialize({ content, serializers });
+  return serializer.deserialize({ content, serializers, preserveId: false });
 };

--- a/packages/apps/types/src/serializer/serializers/default.ts
+++ b/packages/apps/types/src/serializer/serializers/default.ts
@@ -20,8 +20,15 @@ export interface TypedObjectSerializer<T extends Expando = Expando> {
    * @param params.object Deserializing into an existing object. If not provided, a new object is created.
    * @param params.file File metadata.
    * @param params.serializers
+   * @param params.preserveId Preserve the object id. @default true
    */
-  deserialize(params: { content: string; file?: SerializedObject; object?: T; serializers: SerializerMap }): Promise<T>;
+  deserialize(params: {
+    content: string;
+    file?: SerializedObject;
+    object?: T;
+    preserveId?: boolean;
+    serializers: SerializerMap;
+  }): Promise<T>;
 }
 
 export type SerializerMap = Record<string, TypedObjectSerializer>;
@@ -42,23 +49,23 @@ export const jsonSerializer: TypedObjectSerializer = {
     return JSON.stringify(object.toJSON(), null, 2);
   },
 
-  deserialize: async ({ content, object }) => {
+  deserialize: async ({ content, object, preserveId }) => {
     const parsed = JSON.parse(content);
     if (!object) {
-      return deserializeEchoObject(parsed);
+      return deserializeEchoObject(parsed, preserveId);
     } else {
       const { '@id': _, '@type': __, '@meta': ___, ...data } = parsed;
       Object.entries(data)
         .filter(([key]) => !key.startsWith('@'))
         .forEach(([key, value]: any) => {
-          object[key] = isSerializedEchoObject(value) ? deserializeEchoObject(value) : value;
+          object[key] = isSerializedEchoObject(value) ? deserializeEchoObject(value, preserveId) : value;
         });
       return object;
     }
   },
 };
 
-const deserializeEchoObject = (parsed: any): Expando => {
+const deserializeEchoObject = (parsed: any, preserveId = true): Expando => {
   const { '@id': id, '@type': type, '@meta': meta, ...data } = parsed;
   const entries = Object.entries(data)
     .filter(([key]) => !key.startsWith('@'))
@@ -76,8 +83,10 @@ const deserializeEchoObject = (parsed: any): Expando => {
   );
 
   const core = getObjectCore(deserializedObject);
-  core.id = id;
-  getMeta(core).keys.push(...(meta?.keys ?? []));
+  if (preserveId) {
+    core.id = id;
+  }
+  getMeta(deserializedObject).keys.push(...(meta?.keys ?? []));
   const typeRef = getTypeRef(type);
   if (typeRef) {
     core.setType(typeRef);

--- a/packages/apps/types/src/serializer/serializers/default.ts
+++ b/packages/apps/types/src/serializer/serializers/default.ts
@@ -20,13 +20,13 @@ export interface TypedObjectSerializer<T extends Expando = Expando> {
    * @param params.object Deserializing into an existing object. If not provided, a new object is created.
    * @param params.file File metadata.
    * @param params.serializers
-   * @param params.preserveId Preserve the object id. @default true
+   * @param params.newId Generate new ID for deserialized object.
    */
   deserialize(params: {
     content: string;
     file?: SerializedObject;
     object?: T;
-    preserveId?: boolean;
+    newId?: boolean;
     serializers: SerializerMap;
   }): Promise<T>;
 }
@@ -49,29 +49,29 @@ export const jsonSerializer: TypedObjectSerializer = {
     return JSON.stringify(object.toJSON(), null, 2);
   },
 
-  deserialize: async ({ content, object, preserveId }) => {
+  deserialize: async ({ content, object, newId }) => {
     const parsed = JSON.parse(content);
     if (!object) {
-      return deserializeEchoObject(parsed, preserveId);
+      return deserializeEchoObject(parsed, newId);
     } else {
       const { '@id': _, '@type': __, '@meta': ___, ...data } = parsed;
       Object.entries(data)
         .filter(([key]) => !key.startsWith('@'))
         .forEach(([key, value]: any) => {
-          object[key] = isSerializedEchoObject(value) ? deserializeEchoObject(value, preserveId) : value;
+          object[key] = isSerializedEchoObject(value) ? deserializeEchoObject(value, newId) : value;
         });
       return object;
     }
   },
 };
 
-const deserializeEchoObject = (parsed: any, preserveId = true): Expando => {
+const deserializeEchoObject = (parsed: any, newId?: boolean): Expando => {
   const { '@id': id, '@type': type, '@meta': meta, ...data } = parsed;
   const entries = Object.entries(data)
     .filter(([key]) => !key.startsWith('@'))
     .map(([key, value]) => {
       if (isSerializedEchoObject(value)) {
-        return [key, deserializeEchoObject(value)];
+        return [key, deserializeEchoObject(value, newId)];
       }
 
       return [key, value];
@@ -83,7 +83,7 @@ const deserializeEchoObject = (parsed: any, preserveId = true): Expando => {
   );
 
   const core = getObjectCore(deserializedObject);
-  if (preserveId) {
+  if (!newId) {
     core.id = id;
   }
   getMeta(deserializedObject).keys.push(...(meta?.keys ?? []));

--- a/packages/apps/types/src/serializer/serializers/document.ts
+++ b/packages/apps/types/src/serializer/serializers/document.ts
@@ -45,14 +45,14 @@ export const serializer: TypedObjectSerializer<DocumentType> = {
     return `${text}\n\n${footnote}`;
   },
 
-  deserialize: async ({ content, file, object: existingDoc, preserveId }) => {
+  deserialize: async ({ content, file, object: existingDoc, newId }) => {
     if (existingDoc instanceof DocumentType) {
       existingDoc.content!.content = content;
       return existingDoc;
     } else {
       const doc = createEchoObject(create(DocumentType, { content: create(TextV0Type, { content }), comments: [] }));
 
-      if (file && preserveId) {
+      if (file && !newId) {
         const core = getObjectCore(doc);
         core.id = file.id;
       }

--- a/packages/apps/types/src/serializer/serializers/document.ts
+++ b/packages/apps/types/src/serializer/serializers/document.ts
@@ -45,14 +45,14 @@ export const serializer: TypedObjectSerializer<DocumentType> = {
     return `${text}\n\n${footnote}`;
   },
 
-  deserialize: async ({ content, file, object: existingDoc }) => {
+  deserialize: async ({ content, file, object: existingDoc, preserveId }) => {
     if (existingDoc instanceof DocumentType) {
       existingDoc.content!.content = content;
       return existingDoc;
     } else {
       const doc = createEchoObject(create(DocumentType, { content: create(TextV0Type, { content }), comments: [] }));
 
-      if (file) {
+      if (file && preserveId) {
         const core = getObjectCore(doc);
         core.id = file.id;
       }

--- a/packages/apps/types/src/serializer/serializers/index.ts
+++ b/packages/apps/types/src/serializer/serializers/index.ts
@@ -2,14 +2,13 @@
 // Copyright 2024 DXOS.org
 //
 
-// TODO(burdon): Fix import ordering.
 import { type TypedObjectSerializer, jsonSerializer } from './default';
 import { serializer as document } from './document';
-import { serializer as thread } from './thread';
-import { DocumentType, ThreadType } from '../../schema';
+import { serializer as sketch } from './sketch';
+import { DocumentType, SketchType } from '../../schema';
 
 export const serializers: Record<string, TypedObjectSerializer> = {
   [DocumentType.typename]: document,
-  [ThreadType.typename]: thread,
+  [SketchType.typename]: sketch,
   default: jsonSerializer,
 };

--- a/packages/apps/types/src/serializer/serializers/sketch.ts
+++ b/packages/apps/types/src/serializer/serializers/sketch.ts
@@ -15,11 +15,12 @@ export const serializer: TypedObjectSerializer<SketchType> = {
 
   serialize: async ({ object }): Promise<string> => {
     // TODO(wittjosiah): Implement `toJSON` method?
-    const sketch = { ...object, data: { ...object.data } };
+    const data = await loadObjectReferences(object, (s) => s.data);
+    const sketch = { ...object, data: { ...data } };
     return JSON.stringify(sketch, null, 2);
   },
 
-  deserialize: async ({ content, object: existingSketch, preserveId }) => {
+  deserialize: async ({ content, object: existingSketch, newId }) => {
     const parsed = JSON.parse(content);
 
     if (existingSketch instanceof SketchType) {
@@ -31,7 +32,7 @@ export const serializer: TypedObjectSerializer<SketchType> = {
       const data = create(Expando, {});
       const sketch = createEchoObject(create(SketchType, { title: parsed.title, data }));
 
-      if (preserveId) {
+      if (!newId) {
         const core = getObjectCore(sketch);
         core.id = parsed.id;
 

--- a/packages/apps/types/src/serializer/serializers/sketch.ts
+++ b/packages/apps/types/src/serializer/serializers/sketch.ts
@@ -2,8 +2,10 @@
 // Copyright 2024 DXOS.org
 //
 
+import { Expando, loadObjectReferences, createEchoObject, create, getObjectCore } from '@dxos/client/echo';
+
 import { type TypedObjectSerializer, validFilename } from './default';
-import { type SketchType } from '../../schema';
+import { SketchType } from '../../schema';
 
 export const serializer: TypedObjectSerializer<SketchType> = {
   filename: (object) => ({
@@ -12,10 +14,41 @@ export const serializer: TypedObjectSerializer<SketchType> = {
   }),
 
   serialize: async ({ object }): Promise<string> => {
-    throw new Error('Not implemented.');
+    // TODO(wittjosiah): Implement `toJSON` method?
+    const sketch = { ...object, data: { ...object.data } };
+    return JSON.stringify(sketch, null, 2);
   },
 
-  deserialize: async ({ content }) => {
-    throw new Error('Not implemented.');
+  deserialize: async ({ content, object: existingSketch, preserveId }) => {
+    const parsed = JSON.parse(content);
+
+    if (existingSketch instanceof SketchType) {
+      existingSketch.title = parsed.title;
+      const data = await loadObjectReferences(existingSketch, (s) => s.data);
+      setSketchData(data, parsed.data.content);
+      return existingSketch;
+    } else {
+      const data = create(Expando, {});
+      const sketch = createEchoObject(create(SketchType, { title: parsed.title, data }));
+
+      if (preserveId) {
+        const core = getObjectCore(sketch);
+        core.id = parsed.id;
+
+        const dataCore = getObjectCore(data);
+        dataCore.id = parsed.data.id;
+      }
+
+      setSketchData(data, parsed.data.content);
+
+      return sketch;
+    }
   },
+};
+
+const setSketchData = (object: Expando, content: any) => {
+  object.content = {};
+  Object.entries(content).forEach(([key, value]) => {
+    object.content[key] = value;
+  });
 };

